autoload/lsp/buffer.vim

```diff
@@ -41,8 +41,45 @@ export def BufLspServerGet(bnr: number): dict<any>
     return {}
   endif
 
-  # TODO implement logic to compute which server to return
-  return bufnrToServers[bnr][0]
+  if domain == null_string
+    return bufnrToServers[bnr][0]
+  endif
+
+  var SupportedCheckFns = {
+  }
+
+  if !SupportedCheckFns->has_key(domain)
+    # If this happns it is a programming error, and should be fixed in the source code
+    :throw $'Error: ''{domain}'' is not a valid domain'
+    return {}
+  endif
+
+  var SupportedCheckFn = SupportedCheckFns[domain]
+
+  var possibleLSPs: list<dict<any>> = []
+
+  for lspserver in bufnrToServers[bnr]
+    if !SupportedCheckFn(lspserver)
+      continue
+    endif
+
+    possibleLSPs->add(lspserver)
+  endfor
+
+  if possibleLSPs->len() == 0
+    return {}
+  endif
+
+  # LSP server is configured to be a provider for 'domain'
+  for lspserver in possibleLSPs
+    if lspserver.customProviding->has_key(domain)
+        && lspserver.customProviding[domain](lspserver)
+      return lspserver
+    endif
+  endfor
+
+  # Return the first LSP server that supports 'domain'
+  return possibleLSPs[0]
 enddef
```

Comment:
The idea behind the code is to use the first server that actually supports a given action. But also letting the user overriding which language server which should be used by providing a: customProving (⚠️ needs a better name) dictionary with functions that returns boolean if a language server should support a action.